docs/03-intermediate/03-wait-wake-system/handle-reply-wait-wake.md

```diff
@@ -0,0 +1,138 @@
+                .expect("Error in sending a message");
+            debug!("HANDLE: Status::Sent");
+            program.status = Status::Sent;
+            program.msg_ids = (msg_id, msg::id());
+            debug!("HANDLE: WAIT");
+            exec::wait();
+        }
+    }
+    debug!("HANDLE: END");
+}
+
+```
+
+At the beginning, as you may have noticed, the program is in a `Status::Waiting` status, and when a `match` occurs, the code moves to the second variant. The program sends a message, sets the program status to `Status::Sent` and saves the identifiers of the current message and the sent message.  After all this call the function `exec::wait()` function, which pauses the code and adds the current message to the waiting list until `exec::wake(message_id)` is called or the gas runs out. The message identifier is passed to the `exec::wake()` function, which is why `msg::id()` is stored in `program.msg_ids`.
+
+Let's move on to the `handle_reply()` function: 
+
+```rust
+#[no_mangle]
+extern "C" fn handle_reply() {
+    debug!("HANDLE_REPLY");
+    let reply_message: String = msg::load().expect("Unable to decode ");
+    let reply_to = msg::reply_to().expect("Failed to query reply_to data");
+    let program = unsafe { PROGRAM.as_mut().expect("The program is not initialized") };
+
+    if reply_to == program.msg_ids.0 && program.status == Status::Sent {
+        debug!("HANDLE_REPLY: Status::Received");
+        program.status = Status::Received(reply_message);
+        let original_message_id = program.msg_ids.1;
+        debug!("HANDLE: WAKE");
+        exec::wake(original_message_id).expect("Failed to wake message");
+    }
+}
+```
+
+Сondition  `if reply_to == program.msg_ids.0 && program.status == Status::Sent` gives a guarantee that the expected message has arrived and arrived at the right moment, i.e. at the correct program status. 
+After that the status is set to `Status::Received(reply_message)` and the response message is saved; get the ID of the original message and call the `exec::wake()` function, which retrieves the message from the waiting list and the suspended message is restarted in `handle()`. 
+
+*Important note*:  when `exec::wake()` is called, the message is taken from the waiting list, it returns to the `handle()` entrypoint, and the program starts processing it from the beginning, i.e. the program code will get into the `match` again:
+
+```rust
+// ...
+match &program.status {
+    Status::Received(reply_message) => {
+        debug!("HANDLE: Status::Received");
+        msg::reply(reply_message, 0).expect("Error in sending a reply");
+        debug!("HANDLE: Status::Waiting");
+        program.status = Status::Waiting;
+    }
+    // ...
+```
+However, this time it will go into the first variant, send a response and set the status to `Status::Waiting`.
+
+Now, let's examine this process as a whole: 
+
+![Code part 2](../img/03/wait_wake_code.gif)
```

Comment:
Your gifs are AWESOME 🔥 
But on this I'd somehow highlight that actual sending of message will take place only after execution of the message, not in the same millisecond function called